### ISSUE TITLE
Apply pre-commit changes repo-wide to reduce PR chatter and improve contributor/reviewer experience

### ISSUE THREAD

**Issue by ScriptAutomate** [saltstack/salt#59847]
**Description**

Right now, `pre-commit` runs against changes locally to help enforce opinionated standards. This is great for new code updates/changes, but these changes have yet to be applied as baseline formatting against the entire repository.

This leads to a handful of problems. Many files have yet to have:
- `black` applied against them
- `py2` support dropped
- Misc. `pre-commit` config updates around formatting applied

This ends up making each PR, against a previously-untouched file, look like much more is going on than what the initial change was intending. For example, a docstring update to fix a typo could result in 100+ lines of code being auto-formatted by pre-commit, making it difficult for reviewers to parse the actual change (and for users in the future to know what a commit/PR actually had done when looking at git blames).

**Steps to Reproduce the behavior**

- Modify python files that haven't been updated in a while.
- Run pre-commit.
- See all the auto-formatting changes.

Running this against the entire codebase would show the extent of how many files are yet to be modified, which is a significant amount of the codebase.

**Expected behavior**

`pre-commit` should just be fixing/flagging changes a contributor is working on, as the pre-commit baselines should already be applied to the entire codebase where applicable. One a fresh git clone of a repo, pre-commit shouldn't find any failures if ran against the entire repo.

**Comment by ggiesen** [comment]
Given the magnitude of the change (which I fully support), and the suggestion by @s0undt3ch in #59848 that it might be useful to move to the new version of black (20.8b1) but that it would have major impact on the code base, does it make sense to add upgrading black to the list here?

**Comment by danielrobbins** [comment]
@s0undt3ch has explained the current situation to me, regarding maintaining py2 support for the sake of not needing to have to rebase or manually merge several hundred community PR's, and I find this situation unfortunate. Being new to the team, I was not aware of how this worked until now. Essentially, the old version of black and py2 support is being preserved so that existing community PR's don't have to be rebased and can still be merged using a cute little GitHub button.

Historically, when major upstream changes have been made to Salt, this breaks the magical GitHub merge button for existing PRs for Salt, and we have then placed the burden of fixing the PR on our community contributors. I can understand some of them being frustrated when we expect them to do this.

I think the solution to this is just to not ask them to do this and have Salt developers handle manually merging old PR's. Because this is not that hard. And it's also a good way to review the contributions for correctness.

Maintaining py2 compatibility and using an old version of black is really messy, and makes community contribution difficult for the reasons outlined by @ScriptAutomate. Community contributors *need* to use the pre-commit hook to run pyupgrade. This is complicated. It also messes up the commit history because we are mixing Py2->3 updates with community contributions. So we are creating a mess. It's also frankly very weird to have a source tree that is intentionally *not* in compliance with its own CI. That is just not a best practice.

We are essentially penalizing current and future contributions to the Salt project for the sake of our existing PRs. I don't like this. So I support efforts to move things forward. I think we are solving the wrong problem. We should be trying to modernize our code base and make community contributions easier. We should have a code base that is 100% passing its own CI. We can do this without upsetting existing contributors just by doing a tiny bit more work.

**Comment by danielrobbins** [comment]
@ScriptAutomate I have created two issues -- #59853 and #59854 -- which I think can be used as a workaround of sorts for this situation by clearly documenting the current 'interim' setup as we wait for dropping py2 support. This will at least give people the context to understand what is going on now, which should minimize the impact of the current situation.

**Comment by OrangeDog** [comment]
> `black` applied against them

The whole tree was blackened a while ago. That's why there was some trouble with porting from `develop` and at least a few merge conflicts that got resolved incorrectly, introducing new regressions that still aren't all fixed (e.g. #58956).

It's safer to continue as we are now, doing automated changes incrementally.

**Comment by sagetherage** [comment]
Discussed in an internally Salt Core team meeting 2021-MAR-25; we will move forward with communicating this in the weekly Open Hour, starting today to socialize changes across the code base. A PR will be opened next week with these changes and left open for review, a blog post is currently be drafted by the team to be posted in the next 3-4 business days regarding the changes, and we will keep this topic on the weekly Open Hour agenda for the following 2021-APR-01 meeting.

**Comment by waynew** [comment]
I'm +1 for upgrading the newest black (there's a pesky issue around trailing commas, not sure if there are any other problems)

**Comment by danielrobbins** [comment]
@ScriptAutomate I think the big problem is your last point -- we ("we" == typical software developer) expect the tree on GitHub to be in a 'compliant' state, and we expect pre-commit run locally to give us some insight as to whether our code will break the tree or not -- and nothing more. We do not expect the project we are contributing to to intentionally be in a non-compliant state (so, we would expect "pre-commit run --all-files" on a pristine upstream source tree pass successfully. That is not currently the case.)

Additionally, as a contributor to an upstream project, we would typically not expect pre-commit to stealthfully drop py2 support from any files we touch, have this be added to our commit under our name, without our review, and submitted as if it were our personal contribution. We expect it to -- at most -- maybe fix some formatting issues using black. We would not expect our contributions to 'balloon' to include complete rewrites of entire source files. This is confusing and also complicates current PR review, and has some potential copyright implications.

So we are currently subverting these generally accepted expectations of how pre-commit is used. There are two solutions:

1. Keep our pre-commit checks as-is and immediately get the tree compliant with these checks. This will involve immediately running pyupgrade to drop py2 support on all files on GitHub. Maybe we can also upgrade black. Then commits will no longer 'balloon' to include pyupgrade migrations since we have already applied all those changes in our tree. Contributors can then run "pre-commit run --all files" on a pristine tree and our tree should always be in compliance.

2. Drop pyupgrade migrations/checks from the pre-commit check, on both client and CI-side. This is the only area, I think, where our tree is failing pre-commit. Then, it will pass 'pre-commit run --all-files'. So, problem solved. Then at some point in the future we plan to drop py2 support, but do this as ideally a single commit from the Salt Core team, rather than having these pyupgrade changes trickle in as part of various PRs.

The least disruptive solution is number 2.

The more ambitious solution -- with more disruption, is option 1.

I personally would opt for option 2. It is the least disruptive and solves the problem. There isn't really a huge imminent need to drop py2 support. This is just a maintenance task. The code still works with python3 either way.

**Comment by OrangeDog** [comment]
> There are two solutions

I can think of at least two more, and there will be others:

3. Remove pyupgrade from pre-commit and apply it in a separate commit (if necessary) before/after merging a PR.
4. Don't change anything. I don't see why pyupgrade on commit is such a big issue. There's certainly no legal implication. I have no objection or surprise when it does it to me. "we" do not all thing the same way.

Talking about this as "dropping py2 support" is going to be very confusing, as that already happened a few releases ago.

**Comment by waynew** [comment]
I agree with Daniel's premise, my only disagreement is a mild preference for 1.

It *is* disruption, and it *is* some pain, but removing dead code (i.e. six and any related statements) is valuable, and I think worth the disruption and pain. The best time to remove code was when it was no longer used. The second best time is now :slightly_smiling_face: 

This is based on the hopeful assumption that the commits should not cause much conflict. That assumption can be pretty easily tested:

1. Create a new branch
2. Update all of the formatters (black, pyupgrade, etc.)
3. Run against the full tree
4. Run a script that checks out every PR against master and attempts a rebase. Count the success/failures.

Might also be worth also checking a rebase from existing PRs against master.

If there are a ton of failures (maybe more than 25% of PRs against master), that may persuade me that 2 might be the better approach for now. But if there's virtually no change in conflicts between what we have today vs reformatting the whole codebase, I don't see a huge reason to put it off.

**Comment by danielrobbins** [comment]
@OrangeDog Yes, you could theoretically use pre-commit to -- potentially -- send emails to your family if you wanted, because "we" don't do everything the same way, but the first paragraph on the pre-commit home page describes the intended use of the tool, and its intended use does not include how it is being used by Salt currently. Quote (emphasis mine):

*"Git hook scripts are useful for identifying **simple issues before submission to code review**. We run our hooks on every commit to automatically point out issues in code such as **missing semicolons, trailing whitespace, and debug statements**. By pointing these issues out before code review, this allows a code reviewer to **focus on the architecture of a change while not wasting time with trivial style nitpicks.**"*

The Salt use of pre-commit actually **distracts** from the architecture of the change, by combining commits with unrelated changes.

You are also incorrect when you state that there are no potential legal implications of using pre-commit in this way. Any time our git history does not accurately reflect the nature of the actual contribution from a community member then there are potential legal implications. We ideally want the git history to accurately reflect the exact nature of the community contribution. Hopefully you can use your imagination a bit and understand how this could potentially be a problem, without me having to explain various scenarios. It is certainly not a best practice from a copyright and source code traceability perspective to balloon commits in this way.

**Comment by OrangeDog** [comment]
Perhaps you should explain what you think the legal issues are. Simple changes made by an automated process (whether it's changing quote style or removing a six import) do not fall under copyright, regardless of the marked commit author. If you're trying to enumerate the owners of some code for e.g. re-licensing you need to take the list for the whole file anyway, not just the current blame for each line.

**Comment by danielrobbins** [comment]
@OrangeDog  -- Here are the general principles. 

There is no sanity related to copyright claims, and no official process for determining attribution in a legal setting other than having lawyers argue. Anyone can claim there is a copyright issue, and then each side needs to then present an argument/defense, and there is no lack of creativity in these legal arguments, and the motives are generally driven by a desire for financial compensation, not necessarily truth. The best practice to deal with these potential scenarios is to have good documentation. The git history is our authoritative documentation of source code contributions. 

The cleaner the git history, the less ambiguous our attributions, the fewer the complications, and the less legal risk. Conversely, the messier our git history, the more convoluted our attribution of commits, particularly from third parties, the more ambiguous our attributions, and the greater the legal risk. 

Therefore, one should try to avoid all unnecessary complications to our git history, particularly those that are avoidable and attribute wholesale rewrites of core pieces of the source code to various people who did not do it.

Sure, you can make the argument you have stated above in a legal defense, but wouldn't you rather *not* have to make that argument, because you didn't have the problem in the first place? 

Sure, we *could* have git history that combines actual changes with py3 migration rewrites of entire files, but wouldn't it be better if our git history accurately reflect the change they *actually* submitted by user, so we have traceability and good historical documentation on what was actually contributed?

There is a reason why these principles of source code management and traceability exist. They are best practices. We should attempt to follow them.
